flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java

```diff
@@ -174,6 +181,20 @@ public void registerAll(
         }
     }
 
+    @Override
+    public void registerAllAfterRestored(CompletedCheckpoint checkpoint, RestoreMode mode) {
```

Comment:
Because a single completed checkpoint should not be aware of the `highestRetainCheckpointID`; `SharedStateRegistry` would have to have a separate method to update `highestRetainCheckpointID`. 
IMO, that would make the code less readable and more error-prone.

I'm also not sure whether `SharedStateRegistry` should be aware of the restore mode or not;
But given that there are no obvious reasons against it, and taking the above into account, I'd stick with the current approach.

Reply:
Perhaps, I wasn't fully correct in my comment. But I mean that `highestRetainCheckpointID` as field of `SharedStateRegistry` is ok. But instead of `SharedStateRegistry#registerAllAfterRestored(CompletedCheckpoint checkpoint, RestoreMode mode)` we can use following signature `registerAllAfterRestored(CompletedCheckpoint checkpoint, long retainCheckpointID)`. And inside  `CompletedCheckpoint#registerSharedStatesAfterRestored` we can call `registerAllAfterRestored' with the current checkpoint id if the restore mode is not claim and with -1 otherwise.

But I agree that it is a minor change and it is up to you. I just don't really sure that `SharedStateRegistryImpl`should know something about RestoreMode.

Reply:
Thanks for clarifying. Indeed, a separate method would not be needed with this approach. But now the logic inside `CompletedCheckpoint` will be tricky IMO: `registry.registerAll(..., mode == CLAIM ? -1 : this.id`. So I think I'll keep the logic in place in `SharedStateRegistry` for now.